File: survivor-pool-lambda-api/src/main/java/github/jschmidt10/survivor/PoolValidation.java
package github.jschmidt10.survivor;

import github.jschmidt10.survivor.api.Contestant;
import github.jschmidt10.survivor.api.Player;
import github.jschmidt10.survivor.api.Pool;
import github.jschmidt10.survivor.api.Season;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Methods for validating survivor pools.
 */
public class PoolValidation {

    /**
     * Runs all the current validation rules.
     *
     * @param pool
     * @param season
     * @throws RuleViolationException
     */
    public static void validate(Pool pool, Season season) throws RuleViolationException {
        poolIsNamed(pool, season);
        noUnassignedContestants(pool, season);
        noDoubleAssignment(pool, season);
        allPlayersHaveContestants(pool, season);
        allValidContestants(pool, season);
    }

    /**
     * Tests that no contestant is assigned to more than one player.
     *
     * @param pool
     * @param season
     * @throws RuleViolationException
     */
    public static void noDoubleAssignment(Pool pool, Season season) throws RuleViolationException {
        List<Contestant> contestants = pool
                .getPlayers()
                .stream()
                .flatMap((p) -> p.getContestants().stream())
                .collect(Collectors.toList());

        Set<Contestant> distinct = contestants.stream().collect(Collectors.toSet());

        if (distinct.size() != contestants.size()) {
            throw new RuleViolationException("Cannot assign the same castaway to multiple players");
        }
    }

    /**
     * Tests that pool is named.
     *
     * @param pool
     * @param season
     * @throws RuleViolationException
     */
    public static void poolIsNamed(Pool pool, Season season) throws RuleViolationException {
        if (pool.getName() == null) {
            throw new RuleViolationException("You must name your pool!");
        }
    }

    /**
     * Tests that all players have at least one contestant assigned.
     *
     * @param pool
     * @param season
     * @throws RuleViolationException
     */
    public static void allPlayersHaveContestants(Pool pool, Season season) throws RuleViolationException {
        List<Player> emptyPlayers = pool
                .getPlayers()
                .stream()
                .filter((p) -> p.getContestants().isEmpty())
                .collect(Collectors.toList());

        if (!emptyPlayers.isEmpty()) {
            throw new RuleViolationException("Cannot have players with no castaways");
        }
    }

    /**
     * Tests that no contestants are unassigned.
     *
     * @param pool
     * @param season
     * @throws RuleViolationException
     */
    public static void noUnassignedContestants(Pool pool, Season season) throws RuleViolationException {
        Set<Contestant> assignedContestants = pool
                .getPlayers()
                .stream()
                .flatMap((p) -> p.getContestants().stream())
                .collect(Collectors.toSet());

        boolean isUnassigned = season
                .getContestants()
                .stream()
                .anyMatch((c) -> !assignedContestants.contains(c));

        if (isUnassigned) {
            throw new RuleViolationException("Must assign all castaways");
        }
    }

    /**
     * Tests that all assigned contestants are actually in the season.
     *
     * @param pool
     * @param season
     * @throws RuleViolationException
     */
    public static void allValidContestants(Pool pool, Season season) throws RuleViolationException {
        List<Contestant> contestants = pool
                .getPlayers()
                .stream()
                .flatMap((p) -> p.getContestants().stream())
                .collect(Collectors.toList());

        boolean isUnknown = contestants
                .stream()
                .anyMatch((c) -> !season.getContestants().contains(c));

        if (isUnknown) {
            throw new RuleViolationException("Found an unknown castaway");
        }
    }

    /**
     * A rule violation.
     */
    public static class RuleViolationException extends Exception {
        private static final long serialVersionUID = -3501483500806620322L;

        public RuleViolationException(String err) {
            super(err);
        }
    }
}

File: survivorpool-backend/app/pool-validation.js
"use strict";

const assert = require("assert");
const flatmap = require("flatmap");

const ERR_NOT_NAMED = "Pool must be named.";
const ERR_UNASSIGN = "All cast-aways must be assigned.";
const ERR_DOUBLE_ASSIGN = "Cannot assign the same cast-away more than once.";
const ERR_EMPTY_PLAYER = "All players must have at least one cast-away.";
const ERR_INVALID_CONTESTANT = "All cast-aways must be in this season.";

// Possible validation errors
module.exports.ERR_NOT_NAMED = ERR_NOT_NAMED;
module.exports.ERR_UNASSIGN = ERR_UNASSIGN;
module.exports.ERR_DOUBLE_ASSIGN = ERR_DOUBLE_ASSIGN;
module.exports.ERR_EMPTY_PLAYER = ERR_EMPTY_PLAYER;
module.exports.ERR_INVALID_CONTESTANT = ERR_INVALID_CONTESTANT;

/*
 * Validates that a pool is valid. Verifies contestants against the given season.
 */
module.exports.PoolValidator = class PoolValidator {

  /*
   * Validates a pool.
   */
  validate(pool, season) {
    return new Promise((resolve, reject) => {
      try {
        poolIsNamed(pool);
        noUnassignedContestants(pool, season);
        noDoubleAssignments(pool, season);
        allPlayersHaveContestants(pool);
        allValidContestants(pool, season);
        resolve(pool);
      }
      catch (err) {
        reject(err);
      }
    });
  }
};

/*
 * Requires that a pool has a non-empty name.
 */
function poolIsNamed(pool) {
  assert(pool.name && pool.name.length > 0, ERR_NOT_NAMED);
}

/**
 * Validates all the contestants in the season are assigned in the pool.
 */
function noUnassignedContestants(pool, season) {
  var assigned = flatmap(pool.players, (p) => p.contestants.map((c) => c.name));
  season.contestants.forEach((c) => assert(assigned.indexOf(c.name) >= 0, ERR_UNASSIGN));
}

/**
 * Validates no contestant is assigned to multiple players.
 */
function noDoubleAssignments(pool, season) {
  var set = new Set(flatmap(pool.players, (p) => p.contestants.map((c) => c.name)));
  assert(set.size == season.contestants.length, ERR_DOUBLE_ASSIGN);
}

/**
 * Validates all players in the pool have at least one contestant.
 */
function allPlayersHaveContestants(pool) {
  pool.players.forEach((p) => assert(p.contestants.length > 0, ERR_EMPTY_PLAYER));
}

/**
 * Validates all contestants exist in the season.
 */
function allValidContestants(pool, season) {
  var valid = season.contestants.map((c) => c.name);
  var assigned = flatmap(pool.players, (p) => p.contestants);

  assigned.forEach((c) => assert(valid.indexOf(c.name) >= 0, ERR_INVALID_CONTESTANT));
}